src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs

```diff
@@ -0,0 +1,392 @@
+    /// </summary>
+    public int DataRows { get; set;}
+    
+    #endregion
+    
+    #region ITestCaseSource Members
+    
+    /// <summary>
+    /// Returns a set of test case parameters for use as arguments
+    /// to a parameterized test method.
+    /// </summary>
+    /// <param name="method">The method for which data is needed.</param>
+    /// <returns></returns>
+    private IEnumerable<TestCaseParameters> GetTestCasesFor(IMethodInfo method)
+    {
+      List<TestCaseParameters> data = new List<TestCaseParameters>();
+
+      try
+      {
+        IEnumerable source = GetDataFromSource(method);
+
+        if(source != null)
+        {
+          #if NETCF
+          ParameterInfo[] parameters = method.IsGenericMethodDefinition ? new ParameterInfo[0] : method.GetParameters();
+          #else
+          IParameterInfo[] parameters = method.GetParameters();
+          #endif
+
+          foreach (object item in source)
+          {
+            var parms = item as TestCaseParameters;
+
+            if (parms == null)
+            {
+              object[] args = item as object[];
+              if (args != null)
+              {
+                #if NETCF
+                if (method.IsGenericMethodDefinition)
+                {
+                  var mi = method.MakeGenericMethodEx(args);
+                  parameters = mi == null ? new ParameterInfo[0] : mi.GetParameters();
+                }
+                #endif
+                if (args.Length != parameters.Length)
+                  args = new object[] { item };
+              }
+              else if (item is Array)
+              {
+                Array array = item as Array;
+
+                #if NETCF
+                if (array.Rank == 1 && (method.IsGenericMethodDefinition || array.Length == parameters.Length))
+                  #else
+                  if (array.Rank == 1 && array.Length == parameters.Length)
+                    #endif
+                {
+                  args = new object[array.Length];
+                  for (int i = 0; i < array.Length; i++)
+                    args[i] = array.GetValue(i);
+                  #if NETCF
+                  if (method.IsGenericMethodDefinition)
+                  {
+                    var mi = method.MakeGenericMethodEx(args);
+
+                    if (mi == null || array.Length != mi.GetParameters().Length)
+                      args = new object[] {item};
+                  }
+                  #endif
+                }
+                else
+                {
+                  args = new object[] { item };
+                }
+              }
+              else
+              {
+                args = new object[] { item };
+              }
+
+              parms = new TestCaseParameters(args);
+            }
+
+            data.Add(parms);
+          }
+        }
+      }
+      catch (Exception ex)
```

Comment:
The above code seems to duplicate what we have elsewhere to a great extent. Can we refactor to avoid the duplication?